SPIE2016-9913-16/9913-16.tex

```diff
@@ -0,0 +1,303 @@
+We anticipate being able to do a lot by just creating Astropy views to LSST table objects, thereby not doubling the memory requirements, and we have been able to put together a proof-of-concept implementation for doing this.
+This will make LSST tables far more accessible to general users and provide a far richer ecosystem of visualization and debugging tools than would be possible if LSST were required to write their own code.
+
+\subsection{C++ Bindings}
+
+The LSST Data Management team adopted SWIG as the standard method for providing a Python interface to the C++ code.
+Astropy has adopted Cython\cite{2010/content/aip/journal/cise/13/2/10.1109/MCSE.2010.118} when it is required to call C code.
+The Astropy project has no hard rules concerning how C/C++ code should be included in affiliate package but has a strong preference for Cython.
+LSST is not averse to switching away from SWIG (which results in massive code bloat and extended compile times as currently implemented) and has agreed to investigate Cython\cite{dmtn-013} and pybind11\cite{dmtn-014}\footnote{\url{http://pybind11.readthedocs.org/}}.
+One critical requirement is that if we were to move away from SWIG we must be able to do so in an incremental way and so it must be possible for a C++ object produced from SWIG to be passed into a C++ interface created with the alternate approach.
+
+\subsection{Quantities and Units}
+
+With the exception of an angle class, LSST does not use any unit conversion machinery.
+Astropy quantities, a value with an associated unit, are a very powerful class in Astropy that can remove an entire category of bugs from scientific software by ensuring that mathematical operations between quantities use compatible units.
+For heavily used public APIs it would seem to be clear that arguments should be quantities wherever possible.
+The question for LSST is how far down in the stack quantities should be used given the performance hit associated with quantities over plain variables.
+Additionally, how much more complicated does the C++/python interface become if SWIG has to be taught how to convert quantities to the correct form before passing the value to the C++.
+This is especially critical for large arrays, supplied externally, being passed through where we do not want to do a copy operation before entering C++ and also do not want to normalize units for millions of data points at that time.
+Large data arrays and, in particular, table columns returned from C++ layers would benefit from being returned as quantities, as this gives much more flexibility to users of the data and minimizes the risk of misinterpretation of values if units are later changed for a column or the user relies entirely on documentation to interpret values.
+The proposal is that LSST software adopts quantities on a case-by-case basis and are continuously on the look out for  any performance impact associated with quantities.
```

Comment:
"and are continuously" -> ", and its developers will be continuously"?
"look out" -> "lookout"